Improve DateTime parsing

skigrinder opened the pull request:
## Description
- DateTime validation before calling ToInt32() or ToInt64().  Make sure the input string is numeric, otherwise return DateTime.MaxValue

## Motivation and Context
- Prior to these changes, Json deserialization of non-numeric strings longer than 18 characters would cause Native exceptions.  These exceptions caused annoying delays and output when running applications in the nanoFramework debugger.

## How Has This Been Tested?<!-- (IF APPLICABLE) -->
- All unit tests in the Json nuget run successfully in both 'Debug' and 'Release' modes

## Types of changes
- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (work on Unit Tests, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Ellerbach reviewed (commented):
I would make sure we validate performances first. also, I feel like the Regex lib is maybe a bit too much for that. Maybe a simple function to check that all elements are numbers is enough. That may be much faster.
In all cases, I think it's a good idea to try to find the best way to remove the Exception if something else is faster!

josesimoes reviewed (commented):
I'm not confortable, at all, with adding regex validation to this. As @Ellerbach already pointed it's a managed library and the performance it's not the best.

Considering that the code path for numeric deserialization always passes at those points and that this is now making almost 10 calls to regex in order to parse a date time the performance impact its massive!!

There is nothing wrong with the code. Using try catch to catch expected exceptions it's a perfectly valid pattern in C# (and that's why they exist in the first place).

I get it that it can be potencially annoying to see those exceptions showing up on the debugger output. But now that you know they are there, and are harmful to your app, please just ignore them instead of trying to figth them.

If you can come up with a better/improved way of validating a potential datetime value, please, by all means let's have that.

For the sake of this conversation there haven't been any other complaints about this. And there are positive feedback on the current version of the library. This maybe one of those causes that, to paraphrase Mr. Spock in it's Vulcan logic, "the needs of the many outweigh the needs of the few"... 😉

skigrinder commented on nanoFramework.Json.Test/nano.runsettings:
```diff
@@ -9,6 +9,7 @@
    <nanoFrameworkAdapter>
        <Logging>None</Logging>
-       <IsRealHardware>False</IsRealHardware>
+       <IsRealHardware>True</IsRealHardware>
```
Done

skigrinder commented on nanoFramework.Json.Test/nano.runsettings:
```diff
@@ -9,6 +9,7 @@
        <Logging>None</Logging>
-       <IsRealHardware>False</IsRealHardware>
+       <IsRealHardware>True</IsRealHardware>
+       <RealHardwarePort>COM7</RealHardwarePort>
```
Done

skigrinder commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -5,6 +5,7 @@
 //
 
 using System;
+using System.Text.RegularExpressions;
```
Removed.  Please see below

skigrinder commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -63,15 +64,26 @@ public static DateTime FromiCalendar(string iCalendar)
+			// These exceptions cause annoying delays & output when running in the nanoFramework debugger
+			// Bail out if the input string is non-numeric
+			Match match1 = Regex.Match(year, @"^[0-9]*$");
+            Match match2 = Regex.Match(month, @"^[0-9]*$");
```
Removed this code.

skigrinder commented:
As Ellerbach suggested, I've removed the Text.RegularExpressions nuget in favor of a small function to validate a string as numeric.
This causes the code to never call ToInt32() or ToInt64() unless the string is numeric (thereby eliminating the extraneous exceptions).
It also makes the Json library MUCH faster than it was before I started messing with it.
The test case I added (for arrays of class objects) took over 10 seconds before I made any code changes to the Json library.
It took around 9.5 seconds with the Regex code.
It now runs in 2.6 seconds!
It looks like the existing unit test cases are running much faster as well.
Please take a look at the recent changes.

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -109,15 +121,27 @@ public static DateTime FromIso8601(string date)
+			// If the input string is non-numeric, Convert.ToInt32() below will throw an exception from the Native code
+			// These exceptions cause annoying delays & output when running in the nanoFramework debugger
+			// Bail out if the input string is non-numeric
+			Match match1 = Regex.Match(year, @"^[0-9]*$");
```
In case we would use Regex, please also declare the matching as a static compiled element, so we will gain into performance

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -56,13 +56,23 @@ public static DateTime FromiCalendar(string iCalendar)
-			string month = time.Substring(4, 2);
-			string day = time.Substring(6, 2);
-			string hour = time.Substring(9, 2);
+			string year   = time.Substring(0, 4);
```
please us ctrl + k +d to make sure code is properly aligned. The new code is not aligned, has extra spaces, missing some enter.

skigrinder replied:
Done.
Sorry, I learned C in the early 80's and still prefer that style.
Cleaned up a couple other places too.

skigrinder replied:
Just pushed changes to clean the code up.
Please take a look and let me know if I missed anything.
Apparently, my original performance testing must have somehow compared the Release version to the Debug version of the old code (not sure how this happened).

I ran the new code (in Release mode) against the unmodified library (also in Release mode) and the numbers were almost the same.
The numbers are also very close with the same test done in Debug mode.
(Debug mode takes longer for obvious reasons)

Looks like we can't claim a performance benefit with the validation changes in this PR.
But that also means that the Native exceptions are not causing delays.

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -56,13 +56,23 @@ public static DateTime FromiCalendar(string iCalendar)
+			// If the input string is non-numeric, Convert.ToInt32() below will throw an exception from the Native code
+			// These exceptions cause annoying delays & output when running in the nanoFramework debugger
+			// Bail out if the input string is non-numeric
+			if (!IsNumeric(year))   return DateTime.MaxValue;
```
we're not using 1 line if. Please make it multyline with brakets:

```csharp
if (!IsNumeric(year))
{
  DateTime.MaxValue;
}

```

and all } shloud be followed by an extra enter.

skigrinder replied:
Done

Ellerbach reviewed (commented):
I'm still interested in understanding what's the time difference between exception based and this new numerical validation.
I prefer a number than "seems" better for the unit test. If this method is proven to be faster than exception, I'm all in.
Also please make sure you're following the C# coding guidelines. Ctrl+k+d does a lot of magic.  See more elements in the comments.

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -109,15 +137,47 @@ public static DateTime FromIso8601(string date)
 			int milliseconds = Convert.ToInt32(ms);
 
-			if(milliseconds > 999)
+            if (milliseconds > 999)
```
Still some alignment problems in the code. really use Visual Studio 2019, the community version is free and use the magic combination of `ctrl + k + d`

skigrinder replied:
Done.  I've reset my VS options to the default for C# but still not sure if ctrl + k + d is working correctly.

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -109,15 +137,47 @@ public static DateTime FromIso8601(string date)
 			}
 
-			DateTime dt = new(
+            DateTime dt = new(
```
same here for the alignment

skigrinder replied:
Done.  There's no visible difference between the old and new.  Is it correct now?

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -126,7 +186,7 @@ public static DateTime FromIso8601(string date)
                 milliseconds);
 
-			if (utc)
+            if (utc)
```
And here as well

skigrinder replied:
Done. There's no visible difference between the old and new. Is it correct now?

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -179,6 +239,20 @@ public static string ToIso8601(DateTime dt)
+			foreach (char c in str)
+			{
+				if (!((c >= '0') && (c <= '9'))) return false;
+            }
```
And here as well

skigrinder replied:
Done

Ellerbach commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -237,6 +311,18 @@ public static DateTime FromASPNetAjax(string ajax)
+			if (parts.Length < 2)
+			{
+				return DateTime.MaxValue;
+			}
```
you need an extra enter after a closing } exception is for else if in this case

skigrinder replied:
Done.

Ellerbach reviewed (commented):
I'm in favor of using this method rather than the native exception as the performances are the same.
Please correct all the miss alignements.

josesimoes commented on nanoFramework.Json/TimeExtensions.cs:
```diff
@@ -126,178 +186,208 @@ public static DateTime FromIso8601(string date)
+        {
+            foreach (char c in str)
+            {
+                if (!((c >= '0') && (c <= '9')))
```
We probably should have [IsNumber()](https://docs.microsoft.com/en-us/dotnet/api/system.char.isnumber?view=net-5.0) in mscorlib...

But this will work in the meantime.

josesimoes commented:
Here's the performance numbers I have on my machine. (release build)

| Unit Test | Before | New datetime |
|:-|---|---|
| Total | (24 tests) 28ms | (25 tests) 26ms |
| Can_serialize_and_deserialize_arrays_of_class_objects |  --ms | 9ms |
| Can_serialize_and_deserialize_complex_object |  4ms | 4ms |
| CanDeserializeAzureTwinProperties_01 |  <1ms | <1ms |
| CanDeserializeAzureTwinProperties_02 |  6ms | 3ms |
| CanDeserializeAzureTwinProperties_03 |  5ms | 2ms |
| CanDeserializeAzureTwinProperties_04 |  6ms | 2ms |

josesimoes commented:
Last tweaks pushed and then we're good to go!
@skigrinder thanks for the improvements! 💯

skigrinder commented:
Hi Jose' & Laurent,
Thanks for the help getting this implemented!
   Jim

josesimoes commented:
@skigrinder thank YOU for your... determination... 😁  on getting rid of those exception output. In the end, this refactor vastly improved the overall performance. 👍🏻
